### DISCUSSION TITLE
Placeholder resources

### DISCUSSION THREAD

**Discussion by maksloboda** [godotengine/godot-proposals#4779] [Scripting]
Lets say that I want to load some resource on demand.
Common pattern goes like this:
```
func on_some_event():
	print("Resource used")
	var icon : Resource = load("res://icon.png")
	print(icon)
```

![Screenshot from 2022-06-29 17-41-09](https://user-images.githubusercontent.com/26490401/176465127-4c9943dd-890f-48e4-9f1a-eb8d1446a57d.png)

Resource is loaded on demand but we've got a problem, what if the resource moves? In that case our code becomes broken. As a cherry on top this resource is not marked as a dependency, so it can be deleted/moved by an unsuspecting colleague.

Right, but that's what we have export for right?
Lets inspect this code:
```
# Specified in the editor as "res://icon.png"
export var icon : Resource = null

func on_some_event():
	print("Resource used")
	print(icon)
```
![Screenshot from 2022-06-29 17-40-35](https://user-images.githubusercontent.com/26490401/176465824-95f7417c-0c14-440c-bf42-7bee80b56412.png)

Now we have code that tracks the resource path but in return dependency gets loaded when the scene gets loaded.

Okay, but so what? When not using `load` even small projects take a long time to start the game because they need to load nearly everything at once. Level select would have a reference to every level which has a reference to every object and so on.

There exists a feature, that i don't hear about often, called "Load as placeholder". This feature works only for scenes and does almost what i need.
It is possible to create a scene that has an object that contains a Resource. Example:
![Screenshot from 2022-06-29 17-52-27](https://user-images.githubusercontent.com/26490401/176468060-56b4934f-4e29-4c2a-a23e-d14d4fe957a2.png)
But because of the inner mechanics of the child scenes it would require to create and setup separate scene for each resource.

TLDR:
Current resource handling is limited and leads to tradeoff between:
1. Unreliable code
2. Long load times/Excessive RAM use
3. Using weird and hard to setup constructions for a simple action.

Proposed solutions:
1. Ability to do something like `export(String, FILE)` that updates dependencies of the scene and tracks specified resource.
2. class `ProxyResource` that has one method `Resource load()` and allows to create such indirect references.
